### ISSUE TITLE
AssemblyName.Version

### ISSUE THREAD

**Issue by NN---** [dotnet/runtime#37214]
## Background and Motivation

AssemblyName.Version is declared as nullable while documentation doesn't define it as nullable.

**Documentation:**
```cs
public Version Version { get; set; }
```

https://docs.microsoft.com/en-us/dotnet/api/system.reflection.assemblyname.version?view=netcore-3.1

**Property definition:**
```cs
        public Version? Version
        {
            get => _version;
            set => _version = value;
        }

```

https://github.com/dotnet/runtime/blob/master/src/libraries/System.Private.CoreLib/src/System/Reflection/AssemblyName.cs#L44

**Comment by stephentoub** [comment]
The docs in general haven't been updated with nullable annotations yet.

**Comment by NN---** [comment]
I would like to know when it can be null .

**Comment by PathogenDavid** [comment]
TL;DR: It will almost certainly never be null if you're getting it from `Assembly.GetName()`.

It can be null if you're getting processing `AssemblyName` instances. For instance, `new AssemblyName("FakeAssemblyVersion.dll")` results in an `AssemblyName` with a null `Version`.

If you're only working with `Assembly.GetName()`, I'd personally just add an assert to ensure it isn't null in debug builds (because it really won't ever be) and default to 0.0.0.0 in release builds.

-----------------

<details><summary>Click here to read my rambly deepish dive into this topic</summary>
I need a distraction so I went a little too far into this...

The code responsible for initializing the `AssemblyName` object can be seen here:

https://github.com/dotnet/runtime/blob/f62e93416a1799aecc6b0947adad55a0d9870732/src/coreclr/src/vm/assemblyspec.cpp#L428-L434

From what I can find, it'll only be the major/minor version will only be -1 if the assembly identity doesn't have a version:

https://github.com/dotnet/runtime/blob/f9d3e6f2151d090f3fc3eae1965b2339f7a5428e/src/coreclr/src/vm/coreassemblyspec.cpp#L348-L354

A quick glance suggests this flag is always set:

https://github.com/dotnet/runtime/blob/f62e93416a1799aecc6b0947adad55a0d9870732/src/coreclr/src/binder/assemblyname.cpp#L135-L142

I tried both a C# project with no assembly attributes and manually assembling IL to have no version information, and I still got 0.0.0.0 rather than null. I also tried a version of 65535.65535.65535.65535, but the C# compiler errors out and ilasm silently changes the version to 0.0.0.0.

However, what I was able to do is use [Mono.Cecil](https://github.com/jbevain/cecil) to create an assembly with a version of 65535.65535.65535.65535 using the following code: (-1 is casted to a u16 in those code snippets, so this is essentially -1.-1.-1.-1)

```csharp
Version version = new Version(UInt16.MaxValue, UInt16.MaxValue, UInt16.MaxValue, UInt16.MaxValue);
AssemblyNameDefinition assemblyName = new AssemblyNameDefinition("EvilVersionAssembly.dll", version);
AssemblyDefinition assembly = AssemblyDefinition.CreateAssembly(assemblyName, "EvilVersionAssembly", ModuleKind.Dll);
ModuleDefinition module = assembly.MainModule;
TypeDefinition type = new TypeDefinition("EvilVersionAssembly", "Program", TypeAttributes.Public | TypeAttributes.Class, module.TypeSystem.Object);
module.Types.Add(type);
assembly.Write("EvilVersionAssembly.dll");
```

You can't even reference the assembly because the SDK barfs on the null version:

```
1>Microsoft.NET.Sdk.targets(194,5): error MSB4018: The "GenerateDepsFile" task failed unexpectedly.
1>Microsoft.NET.Sdk.targets(194,5): error MSB4018: System.NullReferenceException: Object reference not set to an instance of an object.
1>Microsoft.NET.Sdk.targets(194,5): error MSB4018:    at Microsoft.NET.Build.Tasks.ReferenceInfo.GetVersion(ITaskItem referencePath)
1>Microsoft.NET.Sdk.targets(194,5): error MSB4018:    at Microsoft.NET.Build.Tasks.ReferenceInfo.CreateReferenceInfo(ITaskItem referencePath)
1>Microsoft.NET.Sdk.targets(194,5): error MSB4018:    at Microsoft.NET.Build.Tasks.ReferenceInfo.CreateFilteredReferenceInfos(IEnumerable`1 referencePaths, IEnumerable`1 referenceSatellitePaths)
1>Microsoft.NET.Sdk.targets(194,5): error MSB4018:    at Microsoft.NET.Build.Tasks.ReferenceInfo.CreateDirectReferenceInfos(IEnumerable`1 referencePaths, IEnumerable`1 referenceSatellitePaths, Boolean projectContextHasProjectReferences, Func`2 isRuntimeAssembly)
1>Microsoft.NET.Sdk.targets(194,5): error MSB4018:    at Microsoft.NET.Build.Tasks.GenerateDepsFile.WriteDepsFile(String depsFilePath)
1>Microsoft.NET.Sdk.targets(194,5): error MSB4018:    at Microsoft.NET.Build.Tasks.GenerateDepsFile.ExecuteCore()
1>Microsoft.NET.Sdk.targets(194,5): error MSB4018:    at Microsoft.NET.Build.Tasks.TaskBase.Execute()
1>Microsoft.NET.Sdk.targets(194,5): error MSB4018:    at Microsoft.Build.BackEnd.TaskExecutionHost.Microsoft.Build.BackEnd.ITaskExecutionHost.Execute()
1>Microsoft.NET.Sdk.targets(194,5): error MSB4018:    at Microsoft.Build.BackEnd.TaskBuilder.<ExecuteInstantiatedTask>d__26.MoveNext()
```

However, somehow when I load the assembly using `Assembly.LoadFile` the version comes back as 65535.65535.65535.65535. So there's probably a code path I missed. The reason the SDK goes bang is because it uses the previously mentioned `AssemblyName(string)` constructor:

https://github.com/dotnet/sdk/blob/a5f5bb5183517cb301050288b25c9c727e52e8a6/src/Tasks/Microsoft.NET.Build.Tasks/ReferenceInfo.cs#L150

The assembly name for the Mono.Cecil assembly is `EvilVersionAssembly.dll, Culture=neutral, PublicKeyToken=null`. I suspect that `referencePath.GetMetadata("Version")` call returns `65535.65535.65535.65535` but the `FusionName` one returns the assembly name without a version.

Either which way, as long as Roslyn and ilasm don't produce assemblies without versions and the SDK doesn't accept them as references, you can probably assume your code will never encounter one.
</details>
